File: 프로그래머스/3/42898. 등굣길/등굣길.js
function solution(m, n, puddles) {
    const board = Array.from({length : n+1}, () => Array.from({length:m+1}, () => 0));
    const mod = 1000000007;
    
    for(const puddle of puddles){
        board[puddle[1]][puddle[0]] = -1;
    }
    
    board[1][1] = 1;
    
    for(let i=1; i<n+1; i++) {
        for(let j=1; j<m+1; j++) {
            if(i===1 && j===1) continue;
            if(board[i][j] === -1) continue;
            
            if(board[i-1][j] !== -1) board[i][j] += board[i-1][j];
            if(board[i][j-1] !== -1) board[i][j] += board[i][j-1];
            
            board[i][j] %= mod;
        }
    }    
    
    return board[n][m];
}

File: 프로그래머스/3/42898. 등굣길/등굣길.cpp
#include <string>
#include <vector>
#include <iostream>

using namespace std;

struct Path
{
    long long cnt =0;
    bool isPuddle = false;
};

int solution(int m, int n, vector<vector<int>> puddles) {
    
    // 해당 위치 경로 수
    vector<vector<Path>> board;
    for(int i=0; i<n+1; i++)
    {
        vector<Path> vec(m+1);
        
        board.emplace_back(vec);
    }
    
    // 웅덩이 반영    
    for(auto puddle : puddles)
    {
        board[puddle[1]][puddle[0]].isPuddle = true;
        
        //cout<< puddle[1] << '/' << puddle[0] << '\n';
    }
    
    board[1][1] = Path{1};
    
    for(int i=1; i<n+1; i++)
    {
        for(int j=1; j<m +1; j++) 
        {
            if(i==1 && j==1) continue;
            if(board[i][j].isPuddle) continue;
            
            board[i][j].cnt += board[i][j-1].cnt;
            
            board[i][j].cnt += board[i-1][j].cnt;
            
            board[i][j].cnt %=1000000007;
        }
    }
    
    return board[n][m].cnt;
}